lib/proc/proc_query.cpp

```diff
@@ -628,114 +627,148 @@ namespace {
     bool
     select_parallel(std::shared_ptr<::arrow::internal::ThreadPool> pool) {
       grn_id min_id = get_min_id();
+      // merge isn't parallel. This is just used for queue.
+      auto merge_pool_result = ::arrow::internal::ThreadPool::MakeEternal(1);
+      if (!grnarrow::check(ctx_,
+                           merge_pool_result,
+                           "%s failed to create a thread pool for merging",
+                           tag_)) {
+        return false;
+      }
+      auto merge_pool = *merge_pool_result;
       std::mutex mutex;
       std::vector<grn::UniqueObj> unique_sub_results;
+      auto merge = [&](grn_obj *sub_result) {
+        std::lock_guard<std::mutex> lock(mutex);
+        if (unique_sub_results.empty()) {
+          unique_sub_results.emplace_back(ctx_, sub_result);
+        } else {
+          if (ctx_->rc == GRN_SUCCESS) {
+            grn_table_setoperation(ctx_,
+                                   unique_sub_results[0].get(),
+                                   sub_result,
+                                   unique_sub_results[0].get(),
+                                   GRN_OP_OR);
+          }
+          unique_sub_results.emplace_back(ctx_, sub_result);
+        }
+        return arrow::Status::OK();
+      };
+      std::vector<::arrow::Future<>> merge_futures;
       auto select = [&](grn_obj *match_columns, const std::string &query) {
         auto sub_ctx = grn_ctx_pull_child(ctx_);
+        grn::ChildCtxReleaser releaser(ctx_, sub_ctx);
         grn_obj *condition = build_condition(sub_ctx,
                                              match_columns,
                                              query.data(),
                                              query.length());
+        if (!condition) {
+          return ::arrow::Status::OK();
+        }
+        grn::UniqueObj unique_condition(sub_ctx, condition);
         grn_obj *sub_result = nullptr;
-        if (condition) {
-          grn::UniqueObj unique_condition(sub_ctx, condition);
-          {
-            std::lock_guard<std::mutex> lock(mutex);
-            auto n_unique_sub_results = unique_sub_results.size();
-            if (n_unique_sub_results == 0) {
-              sub_result =
-                grn_table_create(ctx_,
-                                 nullptr, 0,
-                                 nullptr,
-                                 GRN_OBJ_TABLE_HASH_KEY|GRN_OBJ_WITH_SUBREC,
-                                 table_,
-                                 nullptr);
-            } else {
-              sub_result =
-                unique_sub_results[n_unique_sub_results - 1].release();
-              unique_sub_results.pop_back();
-            }
+        {
+          std::lock_guard<std::mutex> lock(mutex);
+          auto n_unique_sub_results = unique_sub_results.size();
+          if (n_unique_sub_results < 2) {
```

Comment:
We're using sized thread pool. Submitted search futures aren't always executed immediately.

If the thread pool has 4 threads and we submit 16 search futures, 12 search futures aren't executed immediately. One of the 12 search futures is executed after one search future is finished.

It means that merge future may be executed zero or more search futures.

Reply:
Ah, does the following block execute the `merge` lambda?
I'm unconfident  the `merge` lambda is executed even if `merge_pool->Submit(merge, sub_result);` failed

```
        auto merge_future = merge_pool->Submit(merge, sub_result);
        if (!merge_future.ok()) {
          std::lock_guard<std::mutex> lock(mutex);
          if (!grnarrow::check(ctx_,
                               merge_future,
                               "%s failed to submit a merge job",
                               tag_)) {
            return merge_future.status();
          }
        }
```

Reply:
It just submits a merge job. The submitted job is executed in another thread. So the job may not be executed in the block.

If `merge_pool->Submit()` is failed, the `merge` lambda isn't executed.

Reply:
Ah, does `Submit()` try to start executing a argument function when it called?
I misunderstood that `Future.status()` try to start executing the function.
`Future.status()` just get the result status, right?

Reply:
> Ah, does `Submit()` try to start executing a argument function when it called?

Right.

> I misunderstood that `Future.status()` tries to start executing the function.
> `Future.status()` just get the result status, right?

No. `Future::status()` waits for finishing (if it's not finished yet) and returns the status.
The `return merge_future.status()` is only executed when `merge_pool->Submit()` is failed. In the case, `merge_future` isn't started, it just returns the (error) status.

Reply:
Successfully created merge futures are waited here: https://github.com/groonga/groonga/pull/1221/files#diff-c9607cf0bb359b7648162fb374b3ce0b4dffa1e4440f8454c933a182c630a778R756

Reply:
> > Ah, does Submit() try to start executing a argument function when it called?

> Right.

Ah, not right. `Submit()` doesn't start execution when the thread pool is full. If the thread pool isn't full, it may start execution immediately in a thread in the thread pool.

Reply:
Thanks for your detailed reply and sorry for my poor expression.
Allow me to explain.

> Ah, does `Submit()` try to start executing a argument function when it's called?

The word `try` tried to express waiting for availability of the thread pool when it is full as you have explained.

> `Future.status()` just get the result status.

The words `just` tried to express `Future.status()` does not try to start the function.
I thought `Future.status()` not just gets result but waits for finishing, I should have written `gets and waits` to avoid misunderstanding...

I'm sorry to bother you...

I understand it, thank you!

Reply:
No problem.
Thanks for your additional explanation.